### ISSUE TITLE
Add "willMakeCredentialWorkWithTheseConstraints()" method to the API

### ISSUE THREAD

**Issue by kpaulh** [w3c/webauthn#345]
We started discussing something like this in the San Francisco f2f meeting. This method would add the ability for an RP to promote the capability to use a built-in authenticator - when and only when the user actually has the appropriate hardware support, without being intrusive.

Very loosely: 
`  Promise<boolean> willMakeCredentialWorkWithTheseConstraints();`

As an example use case, a user logs in to the RP. If the call is successful, the browser would drop an info bar / bubble that informs capable users of the RP's ability to use the authenticator for log-in. The bar provides the option to set it up or defer ("not now") or dismiss all future prompts. 

Choosing the affirmative could create a pop up (or new tab?) that would direct the user through the set-up flow (makeCredential() would be called here).

Alternatives discussed:
- querying the device, which has privacy issues. 
- adding additional bits to makeCredential() that would simply check if the call would *potentially* succeed without actually making a credential. Overloading the method in this way is clunky, might confuse developers.

**Comment by equalsJeffH** [comment]
@kpaulh wrote:
> As an example use case, a user logs in to the RP. If the call is successful..

This is assuming that the login is a legacy one (e.g. via [form-based ("password") authentication](https://en.wikipedia.org/wiki/Form-based_authentication)) not using WebAuthn/FIDO means, yes?
> ..the browser would drop an info bar / bubble that informs capable users of the RP's ability to use the authenticator for log-in..

By "the authenticator" do you mean to say "an available platform authenticator"?  During the discussion at the end of the F2F meeting I recall that roaming authenticators (i.e., those with a cross-platform attachment) were not considered candidates for this use case -- is that still the intention?

Also, this use case depends upon the RP webapp developers to explicitly take advantage of the availability of this method and to weave it properly into their login flows, yes?  If they do not do so, then perhaps there is still need for possibly throwing a "NotFoundError" as anticipated in #302  and #350 ?  @vijaybh ?

**Comment by vijaybh** [comment]
>> As an example use case, a user logs in to the RP. If the call is successful..
> This is assuming that the login is a legacy one (e.g. via form-based ("password") authentication) not using WebAuthn/FIDO means, yes?

I'm not sure that logging in to the RP is even a required part of the use case. It could be that the RP offers this option right on their login page to anonymous users. @kpaulh do you agree?

> By "the authenticator" do you mean to say "an available platform authenticator"? During the discussion at the end of the F2F meeting I recall that roaming authenticators (i.e., those with a cross-platform attachment) were not considered candidates for this use case -- is that still the intention?

That's one possible use case. But the converse is also possible - e.g. a site is trying to figure out whether to upsell you to creating a roaming authenticator for recovery purposes so you can do away with KBA (i.e. the mother's maiden name style questions). Would like to hear from other RPs if they see this as a use case.

> Also, this use case depends upon the RP webapp developers to explicitly take advantage of the availability of this method and to weave it properly into their login flows, yes? If they do not do so, then perhaps there is still need for possibly throwing a "NotFoundError" as anticipated in #302 and #350 ? @vijaybh ?

On reflection, I think these are just orthogonal. This "willMCWork()" is a prelude to the upsell flow, i.e. before deciding to call makeCredential(). Regardless of how well or poorly the RP does things, there is still a desire to not hang the UI (from the user's PoV) when there is no authenticator found after calling makeCredential(). e.g. maybe we're looking for a roaming authenticator, platform says it will work because it knows the user has a USB dongle in their pocket, RP calls makeCredential() but dongle is not inserted.

**Comment by mikewest** [comment]
At the risk of adding noise to the conversation, I'd like to poke a bit at the suggestion that it would be too clunky or confusing to treat this flow as a special-case of `makeCredential`.

It sounds like the proposal at the moment is something like the following, given `userAccountInformation`, `cryptoParams`, `challenge`, and `options`:

```javascript
navigator.authentication
    .promoteAuthenticatorIfAvailable(cryptoParams, options)
    .then(_ => {
      navigator.authentication
        .makeCredential(userAccountInformation, cryptoParams, challenge, options)
        .then(a => {
          // Hooray! We have a thing!
        });
    });
```

That's not a million miles away from:

```javascript
options.checkWithTheUserWhetherThisIsAtAllPossibleFirst = true;
navigator.authentication
  .makeCredential(userAccountInformation, cryptoParams, challenge, options)
  .then(a => {
    // Hooray, we have a thing!
  });
```

In fact, it seems like that's a more direct way of asking for the thing you'd like. If the user clicks through the UI that the user agent provides, you'll end up with a credential. If not, the promise never resolves/rejects/whatever you have in mind.

This falls down a bit if you decide that you want to do something else with the information that a user _could_ create the kind of credential that you're interested in. Navigation would be more difficult, for instance. Is creating a popup or new tab something you expect RPs to want to do?

---

In a credential management world, the latter might be written as:

```javascript
ScopedCredential.create({
  ...,
  checkWithTheUserWhetherThisIsAtAllPossibleFirst: true
})
  .then(c => {
    // Hooray, we have a thing!
  });
```

with the possibility of moving the whole thing to `navigator.credentials.create(...)` depending on the feedback to https://github.com/w3c/webappsec-credential-management/pull/72.

**Comment by vijaybh** [comment]
We discussed that. There are a couple of issues:
1. From a Relying Party UX perspective, sites would typically want to offer friendly guidance along the lines of "We're so glad you want to use Windows Hello! This is what to do next and here is what you should expect to see next time." This might be a popover, navigation, flyout pane, etc.
2. From the UA side, there is no way to tell the difference between "I am calling makeCredential because I know the user has an NFC or USB token in their pocket, pop up a prompt and ask them to tap it or plug it in" and "I'm just probing to see if I should even offer the user a new and more convenient option, so be as quiet as possible and don't annoy them if it's not there".

**Comment by mikewest** [comment]
> From a Relying Party UX perspective, sites would typically want to offer friendly guidance along the lines of "We're so glad you want to use Windows Hello! This is what to do next and here is what you should expect to see next time." This might be a popover, navigation, flyout pane, etc.

Thanks for the clarification, @vijaybh. I guess it's still not clear to me what bits of this flow you expect the user agent to take responsibility for, and what aspects you'd leave up to the site, but as noted below, I'm sold on the separability of the check on the one hand and credential creation on the other.

> From the UA side, there is no way to tell the difference between "I am calling makeCredential because I know the user has an NFC or USB token in their pocket, pop up a prompt and ask them to tap it or plug it in" and "I'm just probing to see if I should even offer the user a new and more convenient option, so be as quiet as possible and don't annoy them if it's not there".

I was suggesting the `checkWithTheUserWhetherThisIsAtAllPossibleFirst` as being that mechanism of informing the UA of the developer's intent. I agree with you that it's the wrong model if you expect the probe to happen in one context, and the `makeCredential` call to happen in another. Is that what you expect? When do you think developers would call this API?

I've been imagining it as part of a sign-up flow, but I realize now that it could also be useful for something like a Google account security checkup flow. In that case you really might not be on the right page (or even right origin) when the developer wants to determine whether the flow would make sense for you.

Thanks! Sorry for the noise. :)

**Comment by vijaybh** [comment]
One clarification - this idea came from @kpaulh and all three browser vendors in that discussion liked it, though maybe we liked it for different reasons. So this is my take, which may be different from what the others think.

In my mind the overall idea is:
- UA takes responsibility for tracking user preferences (e.g. do they ever want to be bothered by any site upselling the plaform authenticator?) and for prompting the user; RP is responsible for orchestrating all flows relating to the credential itself.
- This API is something that could be called indiscriminately by the RP, e.g. whenever it detects that it is running on a host where it has not previously configured a credential on the builtin authenticator. This check and call could perhaps even be in a standard toolbar that the RP always shows at the top of every page. The UA is responsible for converting this to a no-op if the platform does not have a built-in authenticator, or if the user has previously indicated that they don't care to use it for this RP.

I don't think of this as a developer to determine anything. Rather, I think of this as a way for the developer to tell the UA (and hence the user) something along the lines of "here is a thing I can do for you if you're interested, just let me know anytime." There is no expectation that the developer will get a positive (or any) response, but it allows them to present a potentially useful option to the user without fingerprinting the platform and UA in the process. If the user takes them up on the offer, great. If not, no harm done.

So for example you may have logged on to a website with your password and started happily reading your messages, and a little butter bar at the top of the browser tab says did you know you could also set things up so you don't have to enter passwords any more. You click the tell-me-more button, the RP launches a registration flow, and when done returns you to your happy message-reading state. The next time, the logon page just directly asks you to scan your finger, show your face, or whatever else.

I hope that makes things clearer. If not, LMK.

**Comment by AngeloKai** [comment]
This will be addressed by https://github.com/w3c/webauthn/pull/460

**Comment by selfissued** [comment]
This issue was addressed by PR #379 .
